lib/models/user-store.js

```diff
@@ -1,79 +1,231 @@
+      hasMore = connection.pageInfo.hasNextPage;
     }
+
+    this.cache.set(remote, remoteUsers);
   }
 
-  loadUsers() {
-    this.loadUsersFromLocalRepo();
+  addUsers(users, nextSource) {
+    if (
+      nextSource !== this.last.source ||
+      this.repositoryObserver.getActiveModel() !== this.last.repository
+    ) {
+      this.allUsers.clear();
+    }
+
+    let changed = false;
+    for (const author of users) {
+      if (!this.allUsers.has(author.getEmail())) {
+        changed = true;
+      }
+      this.allUsers.set(author.getEmail(), author);
+    }
+
+    if (changed) {
+      this.finalize();
+    }
+    this.last.source = nextSource;
+    this.last.repository = this.repositoryObserver.getActiveModel();
   }
 
-  async loadUsersFromLocalRepo() {
-    const users = await this.repository.getAuthors({max: MAX_COMMITS});
-    const committer = await this.repository.getCommitter();
-    this.setCommitter(committer);
-    this.addUsers(users);
+  finalize() {
+    // TODO: [ku 3/2018] consider sorting based on most recent authors or commit frequency
+    const users = [];
+    for (const author of this.allUsers.values()) {
+      if (author.matches(this.committer)) { continue; }
+      if (author.isNoReply()) { continue; }
+
+      users.push(author);
+    }
+    users.sort(Author.compare);
+    this.users = users;
     this.didUpdate();
   }
 
-  addUsersFromGraphQL(response) {
-    // TODO: [ku 3/2018] also get users from GraphQL API if available. Will need to reshape the data accordingly
-    // This will get called in relay query renderer callback
-    // this.addUsers(users);
+  setRepository(repository) {
+    this.repositoryObserver.setActiveModel(repository);
   }
 
-  addUsers(users) {
-    this.users = {...this.users, ...users};
+  setLoginModel(login) {
+    this.loginObserver.setActiveModel(login);
   }
 
   setCommitter(committer) {
+    const changed = !this.committer.matches(committer);
     this.committer = committer;
+    if (changed) {
+      this.finalize();
+    }
   }
 
   didUpdate() {
-    this.onDidUpdate(this.getUsers());
+    this.emitter.emit('did-update', this.getUsers());
+  }
+
+  onDidUpdate(callback) {
+    return this.emitter.on('did-update', callback);
   }
 
   getUsers() {
-    // TODO: [ku 3/2018] consider sorting based on most recent authors or commit frequency
-    // Also, this is obviously not the most performant. Optimize once we incorporate github username info,
-    // as this will likely impact the shape of the data we store
-    const users = this.users;
-
-    // you wouldn't download a car.  you wouldn't add yourself as a co author.
-    delete users[this.committer.email];
-    delete users[NO_REPLY_GITHUB_EMAIL];
-
-    return Object.keys(users)
-      .map(email => ({email, name: this.users[email]}))
-      .sort((a, b) => {
-        if (a.name < b.name) { return -1; }
-        if (a.name > b.name) { return 1; }
-        return 0;
-      });
+    return this.users;
```

Comment:
question...if users haven't created a GitHub access token, do we fall back on local git history?  Or do we not populate the list of users at all?